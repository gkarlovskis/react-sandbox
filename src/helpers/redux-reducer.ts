import { ReduxActionType } from '../interfaces/enums/redux-action-type';
import { TAppAction } from '../interfaces/types/t-action';
import { TAppState } from '../interfaces/types/t-app-state';

export type DispatchType = (args: TAppAction) => TAppAction;

const initialState: TAppState = {
  isLoggedIn: false,
  user: { username: "" },
};

const reducer = (
  state: TAppState = initialState,
  action: TAppAction
): TAppState => {

  switch (action.type) {
    case ReduxActionType.LOG_IN:
      return {
        ...state,
        isLoggedIn: true,
        user: action.data.user
      };
    case ReduxActionType.LOG_OUT:
      return {
        ...state,
        isLoggedIn: false,
        user: action.data.user
      };
  }
  return state;
};

export default reducer;

