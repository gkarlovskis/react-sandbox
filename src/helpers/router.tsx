import React from 'react';
import { Redirect, Route } from 'react-router-dom';

export const AuthRoute = ({ isAllowed, ...props }: any) => {
  if (isAllowed) {
    return <Route {...props} />;
  } else {
    return <Redirect to="/sign-in" />;
  }
};

export const NonAuthRoute = ({ isLogged, ...props }: any) => {
  if (!isLogged) {
    return <Route {...props} />;
  } else {
    return <Redirect to="/dashboard" />;
  }
};

export const HomeRoute = ({ isLogged, ...props }: any) => {
  if (isLogged) {
    return <Redirect to="/dashboard" />;
  } else {
    return <Redirect to="/sign-in" />;
  }
};
