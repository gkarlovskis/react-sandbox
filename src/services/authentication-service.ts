import { AuthApi } from "./../api/auth-api";
import { IUser } from "./../interfaces/i-user";
import * as dotenvFlow from "dotenv-flow";
dotenvFlow.config({
  path: `${__dirname}/../../config`,
});

export const logout = async () => {
  localStorage.removeItem("currentUser");
  window.location.reload();
  return true;
};

export const login = async (username: string, password: string): Promise<boolean> => {
  const user: IUser = await new AuthApi(
    process.env.AUTH_API_ENDPOINT
      ? process.env.AUTH_API_ENDPOINT
      : "https://postman-echo.com"
  ).fetchUserData(username, password);
  if (!user.username) {
    localStorage.removeItem("currentUser");
    throw new Error("Incorrect password or username");
  }

  // store user details and jwt token in local storage to keep user logged in between page refreshes
  user.passwordDigest = "asdasds1321kljkjklj3123123213jklkx";
  localStorage.setItem("currentUser", JSON.stringify(user));

  return true;
};



export const getLoggedUser = async () => {
  const rawUser = localStorage.getItem("currentUser");
  if (!rawUser) return null;
  try {
    const currentUser: IUser = JSON.parse(rawUser);
    return currentUser ? currentUser : null;
  } catch (error) {
    return null;
  }

};
