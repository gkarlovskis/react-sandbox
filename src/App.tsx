import React, { useEffect, useState } from "react";
// import "../node_modules/bootstrap/dist/css/bootstrap.min.css";
import "./styles/App.css";
import { Router, Switch } from "react-router-dom";
import history from "./helpers/browser-history";
import SignInPage from "./containers/SignInPage";
import SignUpPage from "./containers/SignUpPage";
import Header from "./components/Header";
import Dashboard from "./containers/Dashboard";
import { AppContext } from "./helpers/app-context";
import { getLoggedUser } from "./services/authentication-service";
import { AuthRoute, HomeRoute, NonAuthRoute } from "./helpers/router";
import { useDispatch } from "react-redux";
import { Dispatch } from "redux";
import { IAppState } from "./interfaces/states/i-app-state";
import { IUser } from "./interfaces/i-user";
import { ReduxActionType } from "./interfaces/enums/redux-action-type";

const App: React.FC = () => {
  const [appState, setAppState] = useState<IAppState>({ isLoggedIn: false });

  const dispatch: Dispatch<any> = useDispatch();

  const updateAppState = (user: IUser | null) => {
    if (user === null || user === undefined) {
      dispatch({ type: ReduxActionType.LOG_OUT, data: { isLoggedIn: false } });
      setAppState({ isLoggedIn: false });
    } else {
      dispatch({
        type: user.username !== "" ? ReduxActionType.LOG_IN : ReduxActionType.LOG_OUT,
        data: { user: user },
      });
      setAppState({ isLoggedIn: user.username !== "" ? true : false });
    }
  };

  React.useEffect(() => {
    getLoggedUser().then((user) => {
      updateAppState(user);
    });
  });

  return (
    <AppContext.Provider value={appState}>
      <Router history={history}>
        <Header />
        <Switch>
          <NonAuthRoute isLogged={appState.isLoggedIn} path={"/sign-in"} component={SignInPage} />
          <NonAuthRoute isLogged={appState.isLoggedIn} path={"/sign-up"} component={SignUpPage} />
          <AuthRoute isAllowed={appState.isLoggedIn} exact path="/dashboard" component={Dashboard} />
          <HomeRoute isLogged={appState.isLoggedIn} path={"/"} component={Dashboard} />
        </Switch>
      </Router>
    </AppContext.Provider>
  );
};

export default App;
