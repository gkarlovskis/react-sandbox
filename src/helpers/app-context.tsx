import React from 'react';
import { IAppState } from '../App';

//export const appState: IAppState = { isLoggedIn: false };
//export const AppContext = React.createContext(appState);

export const AppContext = React.createContext<IAppState>({ isLoggedIn: false });

// export class MainPage extends React.Component<undefined, undefined> {
//   static contextType = UserContext;
//   context!: React.ContextType<typeof UserContext>;

//   public render() {
//     return <div>{this.context.isLoggedIn}</div>;
//   }
// }
