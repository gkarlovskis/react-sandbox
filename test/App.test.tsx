import App from "../src/App";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { mount, shallow } from "enzyme";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
const mockStore = configureMockStore();
import { AppContext } from "../src/helpers/app-context";
import { act } from "react-dom/test-utils";

import { getLoggedUser } from "../src/services/authentication-service";
jest.mock("../src/services/authentication-service");

const waitForComponentToPaint = async (wrapper: any) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
    wrapper.update();
  });
};

describe("<App />", () => {
  let wrapper: any;
  const state = { isLoggedIn: false };
  const store: any = mockStore({
    isLoggedIn: true,
  });

  it("Snapshot testing", () => {
    expect(
      shallow(
        <AppContext.Provider value={state}>
          <Provider store={store}>
            <MemoryRouter initialEntries={["/sign-in"]}>
              <Route path="/sign-in">
                <App />
              </Route>
            </MemoryRouter>
          </Provider>
        </AppContext.Provider>
      )
    ).toMatchSnapshot();
  });

  it("App test with logged in state", () => {
    (getLoggedUser as jest.Mock).mockResolvedValue({ username: "mockValidate" });
    wrapper = mount(
      <AppContext.Provider value={state}>
        <Provider store={store}>
          <MemoryRouter initialEntries={["/sign-in"]}>
            <Route path="/sign-in">
              <App />
            </Route>
          </MemoryRouter>
        </Provider>
      </AppContext.Provider>
    );
    waitForComponentToPaint(wrapper);
    const app = wrapper.find("App");
    expect(app.length).toBeGreaterThan(0);
  });

  it("App test with logged out state", () => {
    (getLoggedUser as jest.Mock).mockResolvedValue(null);
    wrapper = mount(
      <AppContext.Provider value={state}>
        <Provider store={store}>
          <MemoryRouter initialEntries={["/sign-in"]}>
            <Route path="/sign-in">
              <App />
            </Route>
          </MemoryRouter>
        </Provider>
      </AppContext.Provider>
    );
    waitForComponentToPaint(wrapper);
    const app = wrapper.find("App");
    expect(app.length).toBeGreaterThan(0);
  });
});
