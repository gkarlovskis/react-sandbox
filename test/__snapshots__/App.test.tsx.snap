// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Snapshot testing 1`] = `ShallowWrapper {}`;
